File: 프로그래머스/lv2/87946. 피로도/피로도.py
import itertools
def solution(k, dungeons):
    MAX = 0
    arr = list(itertools.permutations(dungeons,len(dungeons)))
    for i in range(len(arr)):
        cnt = 0
        num = k
        for j in range(len(dungeons)):
            if num>=arr[i][j][0]:
                cnt+=1
                num-=arr[i][j][1]
            if cnt>MAX:
                MAX = cnt
        num = k
    return MAX

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let max=0
    let visited=Array(dungeons.length).fill(false)
    let cnt=0
    const dfs = (x,cnt) => {
        max=Math.max(cnt,max)
        for(let i=0;i<dungeons.length;i++){
            if(visited[i]) continue
            if(k>=dungeons[i][0]){
                k-=dungeons[i][1]
                visited[i]=true
                cnt++
                dfs(i,cnt)
                k+=dungeons[i][1]
                visited[i]=false
                cnt--
            }
        }
        
    }
    
    for(let i=0;i<dungeons.length;i++){
        if(k>=dungeons[i][0]){
            k-=dungeons[i][1]
            visited[i]=true
            cnt++
            dfs(i,cnt)
            k+=dungeons[i][1]
            visited[i]=false
            cnt--
        }
    }
    
    return max
}